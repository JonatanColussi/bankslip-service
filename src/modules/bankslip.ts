import { BANKSLIP_TYPES } from '../constants';
import { BankslipTypes } from '../types';

export class BankslipBase {
  amount?: number;
  barCode?: string;
  expirationDate?: string;
  type?: BankslipTypes;

  constructor(protected readonly digitableLine: string) {
    this.validateDigitableLine();
    this.setBankslipType();
  }

  protected formatAmount(amount: string) {
    const integer = amount.substring(0, amount.length - 2);
    const cents = amount.slice(-2);

    this.amount = Number.parseFloat(`${integer}.${cents}`) || undefined;
  }

  private setBankslipType() {
    this.type = this.digitableLine.startsWith('8')
      ? BANKSLIP_TYPES.CONVENIO
      : BANKSLIP_TYPES.TITULO;
  }

  private validateDigitableLine() {
    if (!/^\d{47,48}$/.test(this.digitableLine)) {
      throw new Error('The digitable line must contains only numbers');
    }
  }
}
