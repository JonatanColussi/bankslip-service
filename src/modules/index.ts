export * from './bankslip';
export * from './convenio';
export * from './titulo';
