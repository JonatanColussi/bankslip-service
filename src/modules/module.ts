export function module10(block: string) {
  const segments = [...block].map(n => Number(n)).reverse();

  const totalSum = segments.reduce((acc, current, index) => {
    let sum = current * (((index + 1) % 2) + 1);

    sum = sum > 9 ? Math.trunc(sum / 10) + (sum % 10) : sum;

    return acc + sum;
  }, 0);

  return Math.ceil(totalSum / 10) * 10 - totalSum;
}

export function module11(block: string) {
  const segments = [...block].map(n => Number(n)).reverse();

  let multiplier = 2;
  const totalSum = segments.reduce((acc, current) => {
    const sum = current * multiplier;

    multiplier = multiplier === 9 ? 2 : multiplier + 1;

    return acc + sum;
  }, 0);

  const rest = totalSum % 11;

  if ([0, 1].includes(rest)) {
    return 0;
  }

  if (rest === 10) {
    return 1;
  }

  const DV = 11 - rest;

  return DV;
}
