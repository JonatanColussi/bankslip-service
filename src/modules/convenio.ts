import { BankslipBase } from './bankslip';
import { module10, module11 } from './module';

import { BANKSLIP_TYPES } from '../constants';
import { Bankslip } from '../types';

export class Convenio extends BankslipBase implements Bankslip {
  constructor(protected readonly digitableLine: string) {
    super(digitableLine);
    this.validate();
    this.setAmount();
    this.setBarcode();
    this.setExpirationDate();
  }

  setAmount() {
    const reference = Number(this.digitableLine.substring(2, 3));
    const hasAmount = [6, 8].includes(reference);

    if (!hasAmount) {
      this.amount = undefined;

      return;
    }

    const amount = [...this.digitableLine];

    amount.splice(11, 1);

    this.formatAmount(amount.join('').substring(4, 15));
  }

  setBarcode() {
    const barcode = [...this.digitableLine];

    barcode.splice(11, 1);
    barcode.splice(22, 1);
    barcode.splice(33, 1);
    barcode.splice(44, 1);

    this.barCode = barcode.join('');
  }

  setExpirationDate() {
    this.expirationDate = undefined;
  }

  validate() {
    if (this.type !== BANKSLIP_TYPES.CONVENIO) {
      throw new Error('invalid instance type');
    }

    const reference = Number(this.digitableLine.substring(2, 3));
    const module = [6, 7].includes(reference) ? module10 : module11;

    const blocks = [
      {
        block: this.digitableLine.substring(0, 11),
        dv: this.digitableLine.substring(11, 12),
      },
      {
        block: this.digitableLine.substring(12, 23),
        dv: this.digitableLine.substring(23, 24),
      },
      {
        block: this.digitableLine.substring(24, 35),
        dv: this.digitableLine.substring(35, 36),
      },
      {
        block: this.digitableLine.substring(36, 47),
        dv: this.digitableLine.substring(47, 48),
      },
    ];

    if (!blocks.every(item => module(item.block) === Number(item.dv))) {
      throw new Error('The digitable line has invalid checksum');
    }
  }
}
