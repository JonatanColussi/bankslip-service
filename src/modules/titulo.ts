import { BankslipBase } from './bankslip';
import { module10 } from './module';

import { BANKSLIP_TYPES } from '../constants';
import { Bankslip } from '../types';

export class Titulo extends BankslipBase implements Bankslip {
  constructor(protected readonly digitableLine: string) {
    super(digitableLine);
    this.validate();
    this.setAmount();
    this.setBarcode();
    this.setExpirationDate();
  }

  setAmount() {
    const amount = this.digitableLine.substring(37);

    this.formatAmount(amount);
  }

  setBarcode() {
    this.barCode =
      this.digitableLine.substring(0, 4) +
      this.digitableLine.substring(32, 47) +
      this.digitableLine.substring(4, 9) +
      this.digitableLine.substring(10, 20) +
      this.digitableLine.substring(21, 31);
  }

  setExpirationDate() {
    const date = new Date(Date.UTC(1997, 9, 7));
    const days = Number.parseInt(this.digitableLine.substring(33, 37), 10);

    if (Number.isNaN(days) || days === 0) {
      this.expirationDate = undefined;

      return;
    }

    date.setDate(date.getDate() + days);

    const [expirationDate] = date.toISOString().split('T');

    this.expirationDate = expirationDate;
  }

  validate() {
    if (this.type !== BANKSLIP_TYPES.TITULO) {
      throw new Error('invalid instance type');
    }

    const blocks = [
      {
        block: this.digitableLine.substring(0, 9),
        dv: this.digitableLine.substring(9, 10),
      },
      {
        block: this.digitableLine.substring(10, 20),
        dv: this.digitableLine.substring(20, 21),
      },
      {
        block: this.digitableLine.substring(21, 31),
        dv: this.digitableLine.substring(31, 32),
      },
    ];

    if (!blocks.every(item => module10(item.block) === Number(item.dv))) {
      throw new Error('The digitable line has invalid checksum');
    }
  }
}
