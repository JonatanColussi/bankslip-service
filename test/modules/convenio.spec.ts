import { Convenio } from 'modules/convenio';

it('should return an error if intance with invalid bankslip type', () => {
  expect(() => new Convenio('00190000090250038600400684863178390220000048550')).toThrow(
    'invalid instance type',
  );
});

it.each([
  '836200000005667800481800180975657313001589636081',
  '848900000002404201622015809051904292586034111220',
])('should return an error if checksum is invalid', digitableLine => {
  expect(() => new Convenio(digitableLine)).toThrow('The digitable line has invalid checksum');
});

it.each([
  '826100000015165011000008010000870633442022069557',
  '836100000055483500863343933736760093100258416914',
  '846500000001749900791007011198559871921070373342',
  '848900000002404201622015806051904292586034111220',
  '858000000070438403281922630720192528304729600523',
  '817700000000010936599702411310797039001433708318',
])('should parse bankslip properly', digitableLine => {
  expect(new Convenio(digitableLine)).toMatchSnapshot();
});
