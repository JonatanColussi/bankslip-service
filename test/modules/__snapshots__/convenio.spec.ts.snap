// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse bankslip properly 1`] = `
Convenio {
  "amount": 116.5,
  "barCode": "82610000001165011000000100008706344202206955",
  "digitableLine": "826100000015165011000008010000870633442022069557",
  "expirationDate": undefined,
  "type": "CONVENIO",
}
`;

exports[`should parse bankslip properly 2`] = `
Convenio {
  "amount": 548.35,
  "barCode": "83610000005483500863349337367600910025841691",
  "digitableLine": "836100000055483500863343933736760093100258416914",
  "expirationDate": undefined,
  "type": "CONVENIO",
}
`;

exports[`should parse bankslip properly 3`] = `
Convenio {
  "amount": 74.99,
  "barCode": "84650000000749900791000111985598792107037334",
  "digitableLine": "846500000001749900791007011198559871921070373342",
  "expirationDate": undefined,
  "type": "CONVENIO",
}
`;

exports[`should parse bankslip properly 4`] = `
Convenio {
  "amount": 40.42,
  "barCode": "84890000000404201622018060519042958603411122",
  "digitableLine": "848900000002404201622015806051904292586034111220",
  "expirationDate": undefined,
  "type": "CONVENIO",
}
`;

exports[`should parse bankslip properly 5`] = `
Convenio {
  "amount": 743.84,
  "barCode": "85800000007438403281926307201925230472960052",
  "digitableLine": "858000000070438403281922630720192528304729600523",
  "expirationDate": undefined,
  "type": "CONVENIO",
}
`;

exports[`should parse bankslip properly 6`] = `
Convenio {
  "amount": undefined,
  "barCode": "81770000000010936599704113107970300143370831",
  "digitableLine": "817700000000010936599702411310797039001433708318",
  "expirationDate": undefined,
  "type": "CONVENIO",
}
`;
